Allow skipping passed children when restarting a job

Martchus opened the pull request:
* This is particularly useful for directly chained dependencies where it
  does not make sense to restart failed children on-by-one.
* For consistency, expose the ability to skip children generally via the
  API as well. That means it is now possible to skip restarting parents and
  children and only passed children.
* Also improve the error message when restarting is not possible.
* See https://progress.opensuse.org/issues/68956#note-22

Martchus commented:
I've now adapted tests but I'm still sorting out some problems. I've resorted to manual testing because that eases visualizing the dependencies. For easier review, here's a screenshot of a graph which this feature is meant to help with:

![screenshot_20200807_144858](https://user-images.githubusercontent.com/10248953/89647185-6f8ab000-d8bd-11ea-86f0-441776f7fe3c.png)

The user story is basically restarting the failed jobs in this tree of directly chained dependencies without unnecessarily restarting jobs which have passed. So in this case one **wants** to restart `01-child` and `02-child-01…`. One wants to **avoid** restarting `03-child`. In practice there's likely more than just one job like `03-child`. Since the dependencies are directly chained it is also **required** and therefore not avoidable to restart `parent` and `02-child` because those are parents of failed jobs.

When the dependencies would just be regularly chained dependencies one could simply restart the failed children individually. For directly chained dependencies this is not possible because the parents would be cloned multiple times (which is not possible and would be wasteful anyways).

When restarting `parent` all jobs within the tree would be restarted. That behavior is already covering the user story except for the **avoid** part. This change allows to restart `parent` with the `skip_passed_children` flag which would avoid restarting `03-child`. Note that also softfailed jobs would not be restarted.

Martchus commented:
I extended the tests and also the documentation.

Amrysliu commented on docs/WritingTests.asciidoc:
```diff
@@ -329,9 +329,18 @@ have been explicitly defined in a variable for each dependent test suite.
+* When a chained or parallel parent fails or is cancelled, all children will be cancelled.
+* When a parent is restarted, all children are also restarted recursively.
+* When a parallel child is restarted, the parent and siblings will also be restarted.
+* When a *regularly* chained child is restarted, the parent is not restarted. This will usually be fine, but be aware that if an asset uploaded by the chained parent has been cleaned up, the child may fail immediately. To deal with this case, just restart the parent to recreate the asset.
```
These three sentences are the format `when ...,  the ..`, but one of them is `when .., .. will be`, and two of them are `when..., is/are...`, is it better to make them same?

Martchus replied:
I personally think it is consistent enough now.

kalikiana replied:
It reads fine to me. And I think it's nice and easy to understand

Julie-CAO commented:
Hi Marius, restarting all failed children is much helpful for directly chained tests. Thank you very much for coming up with this idea and providing support. About how it works, I'd like to confirm with you.

With this PR, directly chained tests in openQA will work the way below?

-  'restart' button on web UI will restart all the failed children in a chain (as skip_passed_children=1 is the default option when all tests are kicked off for a new build).

    -- Does the original chain will be - re-organized?

    -- Do we have to click the 'restart' button of the parent only? what will happen if I restart 02-child in your test sceenshot? then can we have the chance to rerun 01-child?

- we can run jobs more flexible via REST API with 'skip_passed_children=1', `skip_parents=1` or `skip_children=1`. 

  -- Does the original chain will be re-orgazed by the restart? that determines if a job is allowed to be restarted twice or with different choice.
  
  -- how does REST API work without any of the variables above?

- Does this pr also fix the bug I reported in comment #18 and step 4 in comment #22 step 4 in progress #68956?

Julie-CAO commented:
Hi Marius, restarting all failed children is much helpful for directly chained tests. Thank you very much for coming up with this idea and providing support. About how it works, I'd like to confirm with you.

With this PR, directly chained tests in openQA will work the way below?

-  'restart' button on web UI will restart all the failed children in a chain (as skip_passed_children=1 is the default option when all tests are kicked off for a new build).

    -- Does the original chain will be - re-organized?

    -- Do we have to click the 'restart' button of the parent only? what will happen if I restart 02-child in your test sceenshot? then can we have the chance to rerun 01-child?

- we can run jobs more flexible via REST API with 'skip_passed_children=1', `skip_parents=1` or `skip_children=1`. 

  -- Does the original chain will be re-orgazed by the restart? that determines if a job is allowed to be restarted twice or with different choice.
  
  -- how does REST API work without any of the variables above?

- Does this pr also fix the bug I reported in comment #18 and step 4 in comment #22 step 4 in progress #68956?

Julie-CAO commented:
Hi Marius, restarting all failed children is much helpful for directly chained tests. Thank you very much for coming up with this idea and providing support. About how it works, I'd like to confirm with you.

With this PR, directly chained tests in openQA will work the way below?

-  'restart' button on web UI will restart all the failed children in a chain (as skip_passed_children=1 is the default option when all tests are kicked off for a new build).

    -- Does the original chain will be - re-organized?

    -- Do we have to click the 'restart' button of the parent only? what will happen if I restart 02-child in your test sceenshot? then can we have the chance to rerun 01-child?

- we can run jobs more flexible via REST API with 'skip_passed_children=1', `skip_parents=1` or `skip_children=1`. 

  -- Does the original chain will be re-orgazed by the restart? that determines if a job is allowed to be restarted twice or with different choice.
  
  -- how does REST API work without any of the variables above?

- Does this pr also fix the bug I reported in comment #18 and step 4 in comment #22 step 4 in progress #68956?

Martchus commented:
> 'restart' button on web UI will restart all the failed children in a chain (as skip_passed_children=1 is the default option when all tests are kicked off for a new build).

So far I've kept the default as-is. That means all children will be restarted when restarting the parent (in consistency with other dependency types). So to get the new behavior one needs to use the REST-API with `skip_passed_children=1`. We can of course discuss changing the default.

For the following answers I assume you're talking about `skip_passed_children=1`.

> -- Does the original chain will be - re-organized?

The dependency tree will not be shown for the original jobs because it is generally not show for jobs which have a clone. There's no exception in this case. (Internally no dependency information is ever actually removed from the database, though.) The interesting part is of course how the dependency tree of the skipped jobs will look like. These jobs will be shown within the new graph as shown in the comment above. Not sure whether that counts as "re-organization".

> -- Do we have to click the 'restart' button of the parent only? what will happen if I restart 02-child in your test sceenshot? then can we have the chance to rerun 01-child?

The restart button is available on any level. So you can still restart children. However, that will lead to the parent being cloned as well and that is only possible one time. So restarting individual children only works one time and therefore might not make much sense for directly chained children. Unfortunately it is not easily possible to lift these restrictions because the entire direct cluster needs to be assigned as one unit to the worker and must therefore be created as one unit. That counts for cloned parts of the cluster as well.

> we can run jobs more flexible via REST API with 'skip_passed_children=1', skip_parents=1 or skip_children=1

Yes. I've provided an example how to use the client in the comment above.

> -- Does the original chain will be re-orgazed by the restart? that determines if a job is allowed to be restarted twice or with different choice.

The API behaves identical the the web UI (which also just uses the API). The only difference is the possibility to use these options. Not sure what you mean with re-organized apart from the information I've already given.


> -- how does REST API work without any of the variables above?

It behaves identical the the web UI. None of these options are enabled by default regardless whether you use the web UI or the REST API directly. As mentioned before: We can discuss the defaults. (Although I would not change them as part of this PR anymore.)

> Does this pr also fix the bug I reported in comment #18 and step 4 in comment #22 step 4 in progress #68956?

> step 4: if I run 870 via openqa-clone-job, no error message prompt up and 2 jobs got run. BUT, the restarted 2 jobs behave like two independent jobs completely --- They begin at same time, in no order, and run on different SUT.

That's not fixed. I only amended the restarting behavior. `openqa-clone-job` does not restart jobs (strictly speaking). You can not even use `openqa-clone-job` to initially create a cluster of directly chained dependencies. There's the same problem as with parallel dependencies: The `openqa-clone-job` script creates jobs one-by-one and it is not guaranteed that the jobs will actually be scheduled/assigned/running together.

> Hi Marius, find another problem, not sure if they are the same issue. one child job is 'skipped' when its sibling job is incomplete.

Hard to tell what went wrong here so I'm not sure whether it will be fixed.

---

By the way, I haven't forgotten about your question regarding cancellation:

> When we cancel one child job, the tree with its parent as root got canceled, including its siblings. But we expect only its children and itself got canceled, its siblings should continue. Could this be considered to be fixed as well?

I'm trying to fix this as well as it is clearly wrong to cancel the parent and siblings as well. (One would cancel the parent if that would be wanted.)

kalikiana commented on lib/OpenQA/Resource/Jobs.pm:
```diff
@@ -73,9 +76,14 @@ sub job_restart {
+        }
+        else {
+            push @errors,
+              "It is not possible to restart $job_id. The job (or a dependent job) might have already a clone.";
```
I got slightly confused by the wording. Should it be "...might have been cloned already"? Maybe we could narrow down the reason here? Since the implementation, if I read it correctly, only checks that it was done or cancelled. :thinking:

Martchus replied:
> Since the implementation, if I read it correctly, only checks that it was done or cancelled.

That's true. But the only error which would lead to a negative return value is that single reason. (The error handling should likely still be improved to deal with exceptions. Then I would also include the actual reason.)

kalikiana commented on t/05-scheduler-dependencies.t:
```diff
@@ -1350,6 +1388,58 @@ subtest 'WORKER_CLASS validated when creating directly chained dependencies' =>
+    $_->discard_changes for @all_jobs;
+
+    # duplicate parent
+    $parent->auto_duplicate({skip_passed_children => 1});
```
Maybe add `isnt(undef,...` here to check that we have a clone?

Martchus replied:
That's asserted in the next subtest anyways. As mentioned in the previous comment the error handling should be generally improved (to throw an exception in any error case).

Martchus commented:
@kalikiana Your review comments are basically all concerning the error handling. I've also noticed that the error handling is not nice but would rather fix that in another PR instead of changing too much at once.

kalikiana commented:
> @kalikiana Your review comments are basically all concerning the error handling. I've also noticed that the error handling is not nice but would rather fix that in another PR instead of changing too much at once.

Yeah sounds good. I'm not seeing any mandatory blockers after my second pass.

okurz commented on docs/WritingTests.asciidoc:
```diff
@@ -329,9 +329,18 @@ have been explicitly defined in a variable for each dependent test suite.
+* When a parallel *child* fails or is cancelled, the parent and all other children are also cancelled. This behaviour is intended for closely-related clusters of jobs, e.g. high availability tests, where it's sensible to assume the entire test is invalid if any of its components fails. A special variable can be used to change this behaviour. Setting a parallel parent job's PARALLEL_CANCEL_WHOLE_C ...
+
+====== Further notes
+* The API also allows to skip restarting parents via `skip_parents=1` and to skip restarting children via `skip_children=1`. It is also possible to skip restarting only passed/softfailed children via `skip_passed_children=1`.
```
Saying that `skip_passed_children` also includes softfails is weird. How about `ok_result` as in https://github.com/os-autoinst/openQA/blob/634474bb3d4064b52b636fe57d7fa9e8ba4b6b01/lib/OpenQA/Jobs/Constants.pm#L86 ?

Martchus replied:
So I would call the parameter `skip_ok_result_children`/`skip_ok_children`? I did not chose this because it doesn't sound very nice but maybe it is indeed the better option after all. So should I change the parameter name to one of these?

okurz replied:
I would choose `skip_ok_result_children`. It is explicit and IMHO not too bulky

okurz commented on t/05-scheduler-dependencies.t:
```diff
@@ -1350,6 +1388,58 @@ subtest 'WORKER_CLASS validated when creating directly chained dependencies' =>
+        #       parent is the best behavior but let's assert it for now.
+    } or $log_jobs->() or diag explain $new_job_cluster;
+};
+
```
IMHO t/05-scheduler-dependencies.t became unmaintainable due to being too long now, better change that

Martchus replied:
It is about 1400 lines but a lot of these lines are just big hash/map definitions, graph ASCII arts and some tedious job creations. However, I could easily move out some parts. Like the error handling I would prefer doing the test refactoring as a follow-up. (E.g. the subtest 'serialize sequence of directly chained dependencies' is a good candidate to be moved out but has nothing to do with the rest of the changes of this PR.)

okurz commented on t/api/04-jobs.t:
```diff
@@ -222,15 +222,19 @@ subtest 'job overview' => sub {
+subtest 'restart jobs, error handling' => sub {
+    note 'cluster for 99962';
+    use Data::Dumper;
+    note Dumper($schema->resultset('Jobs')->find(99962)->cluster_jobs);
```
I don't think you need that, explain from Test::Most should be enough, no dumper needed

Martchus replied:
That's just a leftover from debugging. I'll remove it.

okurz commented on templates/webapi/test/infopanel.html.ep:
```diff
@@ -69,7 +69,7 @@
                                 <li>Directly chained parent jobs are restarted.</li>
                             </ul>
-                            All rules are applied recursively. Direct use of the REST-API allows excluding parents.', undef, undef, 'bottom');
+                            All rules are applied recursively. Direct use of the REST-API allows excluding parents and (passed) children.', undef, undef, 'bottom');
```
Also here, not only "passed", better be consistent with the "ok" definition

Julie-CAO commented:
>>>>>>>    'restart' button on web UI will restart all the failed children in a chain (as skip_passed_children=1 is the default option when all tests are kicked off for a new build).
>>>>
>>>>So far I've kept the default as-is. That means all children will be restarted when restarting the parent (in consistency with other dependency types). So to get the new behavior one needs to use the REST-API with skip_passed_children=1. We can of course discuss changing the default.

Julie: sure we need skip_passed_children=1 default for reasons below:

1. REST-API is not as user friendly as web UI, and it needs extra permission.

2. it is very often more than one test in a chain failed. as you described, restart only works one time, then restarting the parent would be the only option to restart these failed tests. That is wasteful and realistic especially for a large tree.

>>>>For the following answers I assume you're talking about skip_passed_children=1.
>>>>
>>>>    -- Does the original chain will be - re-organized?
>>>>
>>>>The dependency tree will not be shown for the original jobs because it is generally not show for jobs which have a clone. There's no exception in this case. (Internally no dependency information is ever actually removed from the database, though.) The interesting part is of course how the dependency tree of the skipped jobs will look like. These jobs will be shown within the new graph as shown in the comment above. Not sure whether that counts as "re-organization".

Julie: 're-organize' here I mean if the original tree is changed after restart. We need keep the test relation to restart them the second time or the third time.

>>>>
>>>>    -- Do we have to click the 'restart' button of the parent only? what will happen if I restart 02-child in your test sceenshot? then can we have the chance to rerun 01-child?
>>>>
>>>>The restart button is available on any level. So you can still restart children. However, that will lead to the parent being cloned as well and that is only possible one time. So restarting individual children only works one time and therefore might not make much sense for directly chained children. Unfortunately it is not easily possible to lift these restrictions because the entire direct cluster needs to be assigned as one unit to the worker and must therefore be created as one unit. That counts for cloned parts of the cluster as well.

Julie: "restarting individual children only works one time", then how many time does 'restarting all failed tests' work? how about next restart if one of the failed test pass in the restart? 
In the case if one of the test is expected to be fail due to a bug, how can we skip it to restart other faied tests? especially it is a long-run test, other tests have to queue until it finishes.


>>>>
>>>>    we can run jobs more flexible via REST API with 'skip_passed_children=1', skip_parents=1 or skip_children=1
>>>>
>>>>Yes. I've provided an example how to use the client in the comment above.
>>>>
>>>>    -- Does the original chain will be re-organized by the restart? that determines if a job is allowed to be restarted twice or with different choice.
>>>>
>>>>The API behaves identical the the web UI (which also just uses the API). The only difference is the possibility to use these options. Not sure what you mean with re-organized apart from the information I've already given.

Julie: my intention is to ask if the failing test is allowed to rerun twice. Allowing a test to be restarted multiple times is necessary for our tests, especially for some unstable tests or unstable environment.

Martchus commented:
> Julie: sure we need skip_passed_children=1 default for reasons below:

The problem with that is that it would change the default for everybody which is likely not wanted. Note that skipping passed children is not specific to any dependency type because I wanted to keep things as simple and generic as possible.

> REST-API is not as user friendly as web UI, […]

That speaks more for allowing to specify these options within the web UI as well.

> […] and it needs extra permission.

One needs to be an operator in both cases. If the web UI allows bypassing required permissions that would be a security issue.

> Julie: 're-organize' here I mean if the original tree is changed after restart.

When I remember the code correctly, the original dependencies are never changed when restarting a job (regardless of this PR and directly chaining). However, the graph is only rendered for the latest jobs in the "cloning chain".

> We need keep the test relation to restart them the second time or the third time.

You can never restart a job which has already been restarted (regardless of this PR and directly chaining).

> Julie: "restarting individual children only works one time", then how many time does 'restarting all failed tests' work? how about next restart if one of the failed test pass in the restart?

I was referring to restarting the individual children of a *concrete* directly chained (or parallel) cluster. Restarting the parent creates a *new* cluster with a new parent and new children. You can restart that *new* parent (and the *newly* failed children).

> In the case if one of the test is expected to be fail due to a bug, how can we skip it to restart other faied tests? especially it is a long-run test, other tests have to queue until it finishes.

You can not skip a particular failed children so far. It would be possible to add a parameter to the API to explicitly list job IDs to be excluded.

> Julie: my intention is to ask if the failing test is allowed to rerun twice. Allowing a test to be restarted multiple times is necessary for our tests, especially for some unstable tests or unstable environment.

Maybe one of my statements was confusing. The general rule about restarting jobs in openQA is: You can not restart a job twice. However, the result of restarting a job is a new job and you can restart that. This creates a "chain" of restarted/cloned jobs and you would always restart the latest job in that chain if you need to restart a job. That's also true for a job which is a directly chained parent.

Martchus commented:
The parameter is now called `skip_ok_result_children` because it is not only about passed (but also softfailed) jobs.

I've created tickets for the outstanding points which came up in the review:

* https://progress.opensuse.org/issues/69868
* https://progress.opensuse.org/issues/69871

grisu48 commented:
Hi there! I'm afraid this update breaks the current workflow in virtualization testing.

What we do

For virtualization, we have the following dependency chain:

![graph_expected](https://user-images.githubusercontent.com/1527829/90105750-bc9bd580-dd46-11ea-8f4f-fee1e461d706.png)

`qam-xen-install` installs the hypervisor and guests and then we run a couple of tests as individual children on this bare-metal hypervisor. Some of the tests occasionally fail and need to be restarted individually. This was working fine until now. Whenever you restarted a job, the dependencies showed up in all jobs.

Now, when you restart a failed child job of depth 1 (e.g. `qam-xen-final`), the job and it's parent get rescheduled. This is unacceptable in our workflow, as it destroys the current state of the hypervisor and thus leaves the whole scenario in a inconsistent state.

In addition, the dependencies of the cloned job are not visible in the openQA web-ui anymore. This also breaks our workflow because we give the original parent job for them as reference and it often happens, that we need to restart children.
Examples: http://openqa.qam.suse.cz/tests/12078 does not show dependencies, it's [clone](http://openqa.qam.suse.cz/tests/12167) does.

![no dependencies shown](https://user-images.githubusercontent.com/1527829/90107110-a68f1480-dd48-11ea-9a57-599099923f26.png)


Also: After restarting a child job, an API call on the restarted parent job only returns the restarted child as child. This breaks usage of [openQA-mon](https://github.com/grisu48/openqa-mon), which we heavily use for monitoring our virtualization runs.
The weird thing is, that the openQA web-ui shows the full dependencies (e.g. http://openqa.qam.suse.cz/tests/12166#dependencies), while a API call on the same job only shows the parent and the child: http://openqa.qam.suse.cz/api/v1/jobs/12166


This is what we expect to see
![oqa_expected](https://user-images.githubusercontent.com/1527829/90107418-12717d00-dd49-11ea-9fdd-8d24fcfabd67.png)


This is what gets shown, because the API call doesn not return the full dependency history anymore
![oqa-shown](https://user-images.githubusercontent.com/1527829/90107413-100f2300-dd49-11ea-8bfb-2db45e4b3185.png)



I understand the idea behind the the logic and the added features would definitely add valuable functionality to openQA, but I'm afraid the default handling has changed in a way that breaks currently the workflow of our virtualization runs which causes major inconveniences.

What we would need in order to get productive again:

* Restarting directly chained children does not by default restarts the parent
* The dependencies of cloned jobs get shown again in the webui and in the job API

For further information don't hesitate to contact me in RocketChat (`ph03nix`)!

Martchus commented:
> Now, when you restart a failed child job of depth 1 (e.g. qam-xen-final), the job and it's parent get rescheduled. This is unacceptable in our workflow, as it destroys the current state of the hypervisor and thus leaves the whole scenario in a inconsistent state.

It was actually this PR which introduced this behavior: https://github.com/os-autoinst/openQA/pull/3262#issue-449595506

And like this PR description states: It is required to restart the parent to ensure both - parent and child - run in sequence on the same worker. If that's not your requirement you should not use a directly chained dependency. For instance, if you simply want to run all your jobs on the same worker, just use a regularly chained dependency and a dedicated worker class. If there's an exceptional case you can still restart the child with `skip_parents=1` (see https://github.com/os-autoinst/openQA/blob/master/docs/WritingTests.asciidoc#further-notes).

> This was working fine until now.

If it was really your requirement that the jobs run directly after each other and on the same worker: Nothing guaranteed that so far because the parent has not been restarted as well. Apparently you have a special worker class and schedule/restart your jobs carefully. If you prefer this manual workflow to ensure the directly chaining you might be better off using regularly chained dependencies.

> In addition, the dependencies of the cloned job are not visible in the openQA web-ui anymore.

That was the case almost from the beginning when I implemented the dependency graph. Not sure why you say "anymore".

> In almost every run we need to restart one or more children.

With directly chained dependencies it is unfortunately not possible to restart only the children as explained. The best workflow for restarting failed children is to restart the parent, possibly via `skip_ok_children=1`.

> What we would need in order to get productive again: Restarting directly chained children does not by default restarts the parent

If you were productive without restarting the parent that means you should better use regularly chained dependencies as this seems to be sufficient in your worker setup.

> The dependencies of cloned jobs get shown again in the webui and in the job API

I implemented that damn graph and I can assure you it never showed for cloned jobs (except for a very short time where that behavior turned out to be problematic). However, I guess we could invest some time to show the graph also for jobs which have a clone.

grisu48 commented:
Thanks for the reply Marius. I'm on to check if switching to regular chained dependencies is a possibility.

> If you were productive without restarting the parent that means you should better use regularly chained dependencies as this seems to be sufficient in your worker setup.
> 
> > The dependencies of cloned jobs get shown again in the webui and in the job API
> 
> I implemented that damn graph and I can assure you it never showed for cloned jobs (except for a very short time where that behavior turned out to be problematic). However, I guess we could invest some time to show the graph also for jobs which have a clone.

You're right! We have always been using the successful parent job as reference and only now, as the parent job also gets restarted this appears to be a problem, where the actual problem lies somewhere else.

Martchus commented:
I've created [a ticket](https://progress.opensuse.org/issues/69976) for showing the dependency graph for cloned jobs. When I remember correctly we didn't invest more time into the dependency graph at the time I implemented it and took the shortcut to hide it for cloned jobs. However, I suppose all the tricky parts mentioned in the issue could be resolved.

Your "I know what I'm doing" setup conflicts with "openQA should do what's generally the right thing". That speaks even more for showing the advanced restarting options also within the web UI so I've also create [a ticket](https://progress.opensuse.org/issues/69979) for that.

grisu48 commented:
> I've created [a ticket](https://progress.opensuse.org/issues/69976) for showing the dependency graph for cloned jobs. When I remember correctly we didn't invest more time into the dependency graph at the time I implemented it and took the shortcut to hide it for cloned jobs. However, I suppose all the tricky parts mentioned in the issue could be resolved.
> 
> Your "I know what I'm doing" setup conflicts with "openQA should do what's generally the right thing". That speaks even more for showing the advanced restarting options also within the web UI so I've also create [a ticket](https://progress.opensuse.org/issues/69979) for that.

Thanks Marius for looking into this. I think the main problem is the new default behaviour for restarting directly chained children. I've moved the discussion to https://github.com/os-autoinst/openQA/pull/3262#issuecomment-673382461

For now, showing the dependency graph for cloned jobs would be a "nice to have", the advanced restarting options are indeed a good idea. I suggested to make the default behaviour of restarting jobs configurable via the config file as this might be something that is not universally the same, but something that might be the same on a instance (as it is in our case) - see https://github.com/os-autoinst/openQA/pull/3262#issuecomment-673382461
